Sources/LeafKit/LeafSyntax/LeafSyntax.swift

```diff
@@ -68,28 +172,110 @@ extension Syntax {
 
         public init(_ params: [ParameterDeclaration], body: [Syntax]) throws {
             guard params.count == 1 else { throw "extend only supports single param \(params)" }
             guard case .parameter(let p) = params[0] else { throw "extend expected parameter type, got \(params[0])" }
             guard case .stringLiteral(let s) = p else { throw "import only supports string literals" }
             self.key = s
+            self.externalsSet = .init(arrayLiteral: self.key)
+            self.importSet = .init()
+            self.exports = [:]
 
-            var exports: [String: Export] = [:]
             try body.forEach { syntax in
                 switch syntax {
                     // extend can ONLY export, raw space in body ignored
                     case .raw: return
                     case .export(let export):
-                        exports[export.key] = export
+                        guard !export.externals().contains(self.key) else {
+                            throw LeafError(.cyclicalReference(self.key, [self.key]))
+                        }
+                        self.exports[export.key] = export
+                        externalsSet.formUnion(export.externals())
+                        importSet.formUnion(export.imports())
                     default:
                         throw "unexpected token in extend body: \(syntax).. use raw space and `export` only"
                 }
             }
+        }
+        
+        internal init(key: String, exports: [String : Syntax.Export], externalsSet: Set<String>, importSet: Set<String>) {
+            self.key = key
             self.exports = exports
+            self.externalsSet = externalsSet
+            self.importSet = importSet
+        }
+        
+        func externals() -> Set<String> {
+            return externalsSet
+        }
+        func imports() -> Set<String> {
+            return importSet
+        }
+        
+        func inlineRefs(_ externals: [String: LeafAST], _ imports: [String : Syntax.Export]) -> [Syntax] {
+            var newExports = [String: Export]()
+            var newImports = imports
+            var newExternalsSet = Set<String>()
+            var newImportSet = Set<String>()
+            
+            // In the case where #exports themselves contain #extends or #imports, rebuild those
+            for (key, value) in exports {
+                guard !value.externals().isEmpty || !value.imports().isEmpty else {
+                    newExports[key] = value
+                    continue
+                }
+                guard case .export(let e) = value.inlineRefs(externals, imports).first else { fatalError() }
+                newExports[key] = e
+                newExternalsSet.formUnion(e.externals())
+                newImportSet.formUnion(e.imports())
+            }
+            
+            // Now add this extend's exports onto the passed imports
+            newExports.forEach {
```

Comment:
This would allow `Dictionary` to resize and merge values in one go: `newExports.merge(newImports, uniquingKeysWith: { $1 })`